src/components/UpgradePurchase/UpgradePurchase.js

```diff
@@ -0,0 +1,95 @@
+import React, { Component } from 'react'
+import classNames from 'classnames'
+import Button from '@material-ui/core/Button'
+import Card from '@material-ui/core/Card'
+import CardHeader from '@material-ui/core/CardHeader'
+import CardActions from '@material-ui/core/CardActions'
+import { array, func, number, object } from 'prop-types'
+
+import { IngredientsList } from '../Recipe/Recipe'
+
+import { canMakeRecipe, doesInventorySpaceRemain } from '../../utils'
+
+import { craftedItems } from '../../img'
+
+import FarmhandContext from '../../Farmhand.context'
+import './UpgradePurchase.sass'
+
+export class UpgradePurchase extends Component {
+  render() {
+    const {
+      handleUpgradeTool,
+      inventory,
+      inventoryLimit,
+      playerInventoryQuantities,
+      upgrade,
+    } = this.props
+
+    const { id, name, description } = upgrade
+
+    const spaceFreedByIngredientsConsumed = Object.values(
+      upgrade.ingredients
+    ).reduce((acc, quantity) => acc + quantity, 0)
+
+    const canBeMade =
+      canMakeRecipe(upgrade, inventory, 1) &&
+      doesInventorySpaceRemain({
+        inventory,
+        // Without the Infinity coercion, this would break recipes for unlimited
+        // inventoryLimits.
+        inventoryLimit:
+          (inventoryLimit === -1 ? Infinity : inventoryLimit) +
+          spaceFreedByIngredientsConsumed,
+      })
```

Comment:
It looks like this was taken from `Recipe.js`: https://github.com/jeremyckahn/farmhand/blob/88cf18b487f351c5eb48d2e669e2efac15fcf53a/src/components/Recipe/Recipe.js#L75-L92

Do you think this could should be refactored out of `Recipe.js` and used in both places? Alternatively, perhaps a TODO to investigate it later?

Reply:
yup! i think putting it somewhere shared makes sense. utils, probably?

Reply:
That makes sense to me!

Reply:
so i went in to actually do this and added a util for getting the total number of ingredients in a recipe, but the `canBeMade` one i backed out of because it ended up looking more confusing with all the args it needs, and it is really just a series of other things that already come from utils. if you still want to break it out we can file an issue for it, but i think it needs further consideration that is more than we need to do for this PR. let me know what you think!

Reply:
Nah that's fine. A little bit of practical repetition is fine and certainly better than inscrutable code. Thanks for giving this a try!